lib/asyn-gai.c

```diff
@@ -0,0 +1,493 @@
+    struct Curl_addrinfo *ca = NULL;
+    CURLcode rc1 = Curl_ai2ca(resolver->gai.cb_buf.ar_result, &ca);
+    CURLcode rc2 = Curl_addrinfo_callback(data, rc1, ca);
+
+#if AGRES_DEBUG
+    infof(data, "RESOLVER: is_resolved: %p done (%d)\n",
+        resolver, rc1|rc2);
+#endif
+
+    if(CURLE_OK != rc1 || CURLE_OK != rc2)
+      return Curl_resolver_error(data);
+
+    *dns = data->state.async.dns;
+    if(!*dns)
+      return Curl_resolver_error(data);
+
+    return CURLE_OK;
+  }
+
+  /* time out if no more time is left for us */
+  resolver->tm.left = Curl_timeleft(data, NULL, 1);
+  if(resolver->tm.left < 0) {
+#if AGRES_DEBUG
+    infof(data, "RESOLVER: is_resolved: %p timeout\n", resolver);
+#endif
+
+    Curl_resolver_cancel(data);
+    Curl_addrinfo_callback(data, CURLE_OPERATION_TIMEDOUT, NULL);
+    return CURLE_OPERATION_TIMEDOUT;
+  }
+
+  /* initialize expire ms for get_sock() and back off for repeated calls */
+  if(resolver->tm.expire == 0)
+    resolver->tm.expire = AGRES_EXPIRE_INIT;
+  else
+    resolver->tm.expire *= AGRES_EXPIRE_GROW;
+  if(resolver->tm.expire > resolver->tm.left)
+    resolver->tm.expire = resolver->tm.left;
+  if(resolver->tm.expire > AGRES_EXPIRE_MAX)
+    resolver->tm.expire = AGRES_EXPIRE_MAX;
+
+  return CURLE_OK;
+}
+
+/*
+ * Curl_resolver_wait_resolv()
+ *
+ * Waits for a resolve to finish. This function should be avoided since using
+ * this risk getting the multi interface to "hang".
+ *
+ * If 'entry' is non-NULL, make it point to the resolved dns entry
+ *
+ * Returns CURLE_COULDNT_RESOLVE_HOST if the host was not resolved,
+ * CURLE_OPERATION_TIMEDOUT if a time-out occurred, or other errors.
+ */
+CURLcode Curl_resolver_wait_resolv(struct Curl_easy *data,
+                                   struct Curl_dns_entry **dnsentry)
+{
+  struct Curl_dns_entry *dns = NULL;
+  CURLcode rc;
+
+#if AGRES_DEBUG
+  infof(data, "RESOLVER: wait: %p\n", data->state.async.resolver);
+#endif
+
+  do {
+    rc = Curl_resolver_is_resolved(data, &dns);
+  } while(CURLE_OK == rc && !dns);
+
+  if(dnsentry)
+    *dnsentry = dns;
+
+  return rc;
+}
+
+/*
+ * Curl_resolver_getaddrinfo() - when using this resolver
+ *
+ * Returns name information about the given hostname and port number. If
+ * successful, the 'hostent' is returned and the forth argument will point to
+ * memory we need to free after use. That memory *MUST* be freed with
+ * Curl_freeaddrinfo(), nothing else.
+ *
+ * Each resolver backend must of course make sure to return data in the
+ * correct format to comply with this.
+ */
+struct Curl_addrinfo *Curl_resolver_getaddrinfo(struct Curl_easy *data,
+                                                const char *hostname,
+                                                int port,
+                                                int *waitp)
+{
+  int gai_rc, pf = PF_INET;
+  struct Curl_async *async = &data->state.async;
+  struct resolver *resolver = async->resolver;
+  const char *error = "unknown error";
```

Comment:
This error message can never be shown, can it?

Reply:
Yep, probably a refactoring left-over.